pkg/controllers/allocation/batch.go

```diff
@@ -0,0 +1,150 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocation
+
+import (
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	BatchCacheTTL             = 1 * time.Hour
+	BatchCacheCleanupInterval = 2 * time.Hour
+)
+
+type Batcher struct {
+	// MaxBatchPeriod is the maximum amount of time to batch incoming pods before flushing
+	MaxBatchPeriod time.Duration
+	// IdlePeriod is the amount of time to wait to flush a batch when there are no incoming pods but the batch is not empty
+	// It should be a smaller duration than MaxBatchPeriod
+	IdlePeriod time.Duration
+
+	// batches keeps a mapping of a key (like a provisioner name and namespace) to a specific Batch
+	batches *cache.Cache
+}
+
+// NewBatcher creates a new batch manager to start multiple batch windows
+func NewBatcher(maxBatchPeriod time.Duration, idlePeriod time.Duration) *Batcher {
+	batchCache := cache.New(BatchCacheTTL, BatchCacheCleanupInterval)
+	batchCache.OnEvicted(func(key string, val interface{}) {
+		batch := val.(*Batch)
+		batch.close()
+	})
+	return &Batcher{
+		MaxBatchPeriod: maxBatchPeriod,
+		IdlePeriod:     idlePeriod,
+		batches:        batchCache,
+	}
+}
+
+// Add starts a batching window or updates an existing one based on a batching key
+func (m *Batcher) Add(key string) {
+	batch, ok := m.batches.Get(key)
+	if !ok {
+		batch = &Batch{
+			Batcher: m,
+			start:   make(chan bool, 1),
+			updates: make(chan bool, 1),
+			end:     make(chan bool, 1),
+		}
+		m.batches.SetDefault(key, batch)
+	}
+	// Updates expiration
+	m.batches.SetDefault(key, batch)
+	batch.(*Batch).Add()
+}
+
+// Complete blocks until a specific batching window ends based on the batching key
+func (m *Batcher) Complete(key string) {
+	batch, ok := m.batches.Get(key)
+	if !ok {
+		return
+	}
+	batch.(*Batch).Complete()
+}
+
+// Batch implements a single batching window based on a max timeout and a progress period
+type Batch struct {
+	// start is a channel to coordinate window starts
+	start chan bool
+	// updates is a channel to communicate progress in filling a batch window
+	updates chan bool
+	// end is a channel to signal a window close
+	end chan bool
+
+	sync.Mutex
+	*Batcher
+}
+
+// Add starts a batching window or adds to an existing in-progress window
+func (b *Batch) Add() {
+	b.Lock()
```

Comment:
Are there some higher-level abstractions you can use here instead of raw Lock? (WaitGroups or something else?) I haven't really thought it true but this kind of thing can be deceptively tricky.

Reply:
I could use a channel (first draft had a channel) or a WaitGroup, but honestly I think the lock in this case is the easiest to reason about. The channel seemed a bit convoluted to think and look at.

Reply:
See my other comment around how to simplify this - I'm not sure any synchronization is really needed.